File: commands/stats/contest_streak.py
# https://github.com/viktorexe/terristats-discord-bot
import discord
from discord.ext import commands
from discord import app_commands
from datetime import datetime, timezone
import re
from database import db

class ContestStreak(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @app_commands.command(name='contest_streak', description='Show clan contest win streaks')
    async def contest_streak(self, interaction: discord.Interaction, clan: str, hours: int = None):
        try:
            await interaction.response.defer()
            
            if hours is not None and (hours <= 0 or hours > 720):
                await interaction.followup.send(embed=discord.Embed(
                    title="❌ Invalid Hours", 
                    description="Hours must be between 1 and 720 (30 days).", 
                    color=0xff0000
                ), ephemeral=True)
                return
            
            clan_upper = clan.upper().strip()
            query = {'clan_name': {'$regex': f'^{re.escape(clan_upper)}$', '$options': 'i'}}
            
            if hours is not None:
                from datetime import timedelta
                cutoff_time = datetime.now(timezone.utc) - timedelta(hours=hours)
                query['timestamp'] = {'$gte': cutoff_time}
                time_period = f"Last {hours} Hours"
            else:
                time_period = "All Time"
            
            pipeline = [
                {'$match': query},
                {'$match': {'contest': True}},
                {'$sort': {'timestamp': 1}},
                {'$project': {'timestamp': 1}}
            ]
            
            contest_only_wins = await db.db.clanwins.aggregate(pipeline).to_list(None)
            
            if not contest_only_wins:
                total_contests = 0
            else:
                total_contests = len(contest_only_wins)
                
                temp_streak = 1
                best_streak = 1
                current_streak = 1
                best_streak_start = contest_only_wins[0].get('timestamp')
                
                for i in range(1, len(contest_only_wins)):
                    prev_time = contest_only_wins[i-1].get('timestamp', datetime.min.replace(tzinfo=timezone.utc))
                    curr_time = contest_only_wins[i].get('timestamp', datetime.min.replace(tzinfo=timezone.utc))
                    
                    if prev_time.tzinfo is None:
                        prev_time = prev_time.replace(tzinfo=timezone.utc)
                    if curr_time.tzinfo is None:
                        curr_time = curr_time.replace(tzinfo=timezone.utc)
                    
                    time_diff = (curr_time - prev_time).total_seconds() / 60  # Convert to minutes
                    
                    if time_diff <= 20:
                        temp_streak += 1
                        if temp_streak > best_streak:
                            best_streak = temp_streak
                            best_streak_start = contest_only_wins[i - temp_streak + 1].get('timestamp')
                    else:
                        temp_streak = 1
                
                current_streak = temp_streak
            
            embed = discord.Embed(
                title=f"[{clan_upper}] Contest Streaks",
                description=f"{time_period}\nTotal contest wins: {total_contests}",
                color=0x2b2d31
            )
            
            if current_streak > 0:
                embed.add_field(
                    name="🔥 Current Streak",
                    value=f"**{current_streak}** consecutive contest wins",
                    inline=False
                )
            else:
                embed.add_field(
                    name="🔥 Current Streak",
                    value="No contest wins in time period",
                    inline=False
                )
            
            if best_streak > 0:
                date_str = best_streak_start.strftime('%b %d, %Y') if best_streak_start else 'Unknown'
                embed.add_field(
                    name="🏆 Best Streak",
                    value=f"**{best_streak}** consecutive contest wins\nStarted: {date_str}",
                    inline=False
                )
            else:
                embed.add_field(
                    name="🏆 Best Streak",
                    value="No contest wins recorded",
                    inline=False
                )
            
            total_games = await db.db.clanwins.count_documents({})
            embed.set_footer(text=f"Data from 18 Nov 2025 | {total_games:,} games tracked")
            await interaction.followup.send(embed=embed)
            
        except Exception as e:
            print(f"Error in contest_streak: {e}")
            try:
                await interaction.followup.send(embed=discord.Embed(
                    title="Error", description="Failed to calculate contest streak.", color=0xff0000
                ))
            except:
                pass

async def setup(bot):
    await bot.add_cog(ContestStreak(bot))

File: commands/stats/contest_streak.js
// https://github.com/viktorexe/terristats-discord-bot
// /contest_streak - current and best consecutive contest-win streaks for a
// clan. A streak continues while consecutive contest wins are <= 20 minutes
// apart.
const { SlashCommandBuilder, EmbedBuilder } = require('discord.js');
const db = require('../../database');
const { exactMatch, gamesFooter } = require('../../utils');

// Format a Date like Python's "%b %d, %Y", e.g. "Nov 18, 2025".
function formatDate(date) {
  if (!date) return 'Unknown';
  const months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
  const d = new Date(date);
  return `${months[d.getUTCMonth()]} ${String(d.getUTCDate()).padStart(2, '0')}, ${d.getUTCFullYear()}`;
}

module.exports = {
  data: new SlashCommandBuilder()
    .setName('contest_streak')
    .setDescription('Show clan contest win streaks')
    .addStringOption((o) =>
      o.setName('clan').setDescription('Clan name').setRequired(true),
    )
    .addIntegerOption((o) =>
      o.setName('hours').setDescription('Look-back window (1-720)').setRequired(false),
    ),

  async execute(interaction) {
    try {
      await interaction.deferReply();

      const clan = interaction.options.getString('clan');
      const hours = interaction.options.getInteger('hours');

      if (hours !== null && (hours <= 0 || hours > 720)) {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle('❌ Invalid Hours')
              .setDescription('Hours must be between 1 and 720 (30 days).')
              .setColor(0xff0000),
          ],
        });
        return;
      }

      const clanUpper = clan.toUpperCase().trim();
      const query = { clan_name: exactMatch(clanUpper) };

      let timePeriod;
      if (hours !== null) {
        query.timestamp = { $gte: new Date(Date.now() - hours * 3600 * 1000) };
        timePeriod = `Last ${hours} Hours`;
      } else {
        timePeriod = 'All Time';
      }

      const contestWins = await db
        .collection('clanwins')
        .aggregate([
          { $match: query },
          { $match: { contest: true } },
          { $sort: { timestamp: 1 } },
          { $project: { timestamp: 1 } },
        ])
        .toArray();

      let totalContests = 0;
      let currentStreak = 0;
      let bestStreak = 0;
      let bestStreakStart = null;

      if (contestWins.length) {
        totalContests = contestWins.length;
        let tempStreak = 1;
        bestStreak = 1;
        bestStreakStart = contestWins[0].timestamp;

        for (let i = 1; i < contestWins.length; i += 1) {
          const prev = new Date(contestWins[i - 1].timestamp);
          const curr = new Date(contestWins[i].timestamp);
          const diffMinutes = (curr - prev) / 60000;

          if (diffMinutes <= 20) {
            tempStreak += 1;
            if (tempStreak > bestStreak) {
              bestStreak = tempStreak;
              bestStreakStart = contestWins[i - tempStreak + 1].timestamp;
            }
          } else {
            tempStreak = 1;
          }
        }
        currentStreak = tempStreak;
      }

      const embed = new EmbedBuilder()
        .setTitle(`[${clanUpper}] Contest Streaks`)
        .setDescription(`${timePeriod}\nTotal contest wins: ${totalContests}`)
        .setColor(0x2b2d31);

      embed.addFields({
        name: '🔥 Current Streak',
        value:
          currentStreak > 0
            ? `**${currentStreak}** consecutive contest wins`
            : 'No contest wins in time period',
        inline: false,
      });

      embed.addFields({
        name: '🏆 Best Streak',
        value:
          bestStreak > 0
            ? `**${bestStreak}** consecutive contest wins\nStarted: ${formatDate(bestStreakStart)}`
            : 'No contest wins recorded',
        inline: false,
      });

      const totalGames = await db.collection('clanwins').countDocuments({});
      embed.setFooter({ text: gamesFooter(totalGames) });
      await interaction.editReply({ embeds: [embed] });
    } catch (err) {
      console.log(`Error in contest_streak: ${err.message}`);
      try {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle('Error')
              .setDescription('Failed to calculate contest streak.')
              .setColor(0xff0000),
          ],
        });
      } catch (e) {
        // ignore
      }
    }
  },
};
